### ISSUE TITLE
Incorrect `invalid-return-type` in type-preserving generic mapping function branching on runtime type

### ISSUE THREAD

**Issue by shilch** [astral-sh/ty#483]
### Summary

Following function is flagged incorrectly(?) my ty:

```python
def func[T](val: T) -> T:
    if type(val) is str:
        return ''
    return val
```

The following output is generated:

```
error[invalid-return-type]: Return type does not match returned value
 --> test.py:1:24
  |
1 | def func[T](val: T) -> T:
  |                        - Expected `T` because of return type
2 |     if type(val) is str:
3 |         return ''
  |                ^^ expected `T`, found `Literal[""]`
4 |     return val
  |
info: rule `invalid-return-type` is enabled by default
```

This behavior is consistent with mypy but looks erroneous: https://github.com/python/mypy/issues/12989#issuecomment-1691226161

### Version

ty 0.0.1-alpha.6

**Comment by carljm** [comment]
Thanks for the report! Our behavior here is consistent with both mypy and pyright.

I think the requested behavior here is safe only if you do a "type(...) is" check (not e.g. an `isinstance` check which would permit subclasses). (This means it would also require that we add support for "exact" instance types which exclude subclasses.)

It also requires ensuring that a generic function typevar can never be solved to a more precise type than what can be determined at runtime via `type(...) is`. For example, a typevar can never be solved to a literal type (currently we will). This also wouldn't work in any case involving a generic type instead of `str`, since generics are erased at runtime and thus can't be checked via `type(...) is`.

All in all, I don't think it's out of the question that we might someday consider this, but I'm not sure we will ever support it, and it's certainly low priority for now.
